libraries/AP_Hott_Telem/AP_Hott_Telem.cpp

```diff
@@ -0,0 +1,358 @@
+        msg.gps_speed_kmh = uint16_t(gps.ground_speed() * 3.6 + 0.5);
+        float sacc, hacc;
+        if (gps.speed_accuracy(sacc)) {
+            msg.speed_acc = sacc * 100 + 0.5;
+        }
+        if (gps.horizontal_accuracy(hacc)) {
+            msg.horiz_acc = hacc * 100 + 0.5;
+        }
+    }
+
+    float lat = loc.lat * 1.0e-7;
+    float lon = loc.lng * 1.0e-7;
+
+    uint16_t dm, sec;
+    GPS_to_DMS(lat, msg.pos_NS, dm, sec);
+    msg.pos_NS_dm = dm;
+    msg.pos_NS_sec = sec;
+
+    GPS_to_DMS(lon, msg.pos_EW, dm, sec);
+    msg.pos_EW_dm = dm;
+    msg.pos_EW_sec = sec;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    Vector2f home_vec;
+    float posD = 0;
+    {
+        WITH_SEMAPHORE(ahrs.get_semaphore());
+        if (ahrs.get_relative_position_NE_home(home_vec)) {
+            msg.home_distance = home_vec.length();
+        }
+        ahrs.get_relative_position_D_home(posD);
+        Vector3f vel;
+        if (ahrs.get_velocity_NED(vel)) {
+            msg.climbrate = uint16_t(30000.5 + vel.z * -100);
+            msg.climbrate3s = 120 + vel.z * -3;
+            msg.vel_north = vel.x * 1000 + 0.5;
+            msg.vel_east = vel.y * 1000 + 0.5;
+        }
+    }
+
+    msg.altitude = uint16_t(500.5 - posD);
+
+    msg.gps_satelites = gps.num_sats();
+    switch (gps.status()) {
+    case AP_GPS::NO_GPS:
+    case AP_GPS::NO_FIX:
+        msg.gps_fix_char = '-';
+        break;
+    case AP_GPS::GPS_OK_FIX_2D:
+        msg.gps_fix_char = '2';
+        break;
+    default:
+        msg.gps_fix_char = '3';
+        break;
+    }
+    msg.gps_fix_char2 = msg.gps_fix_char;
+
+    msg.home_direction = degrees(atan2f(home_vec.y, home_vec.x)) * 0.5 + 0.5;
+
+    AP_RTC &rtc = AP::rtc();
+    {
+        WITH_SEMAPHORE(rtc.get_semaphore());
+        uint16_t ms;
+        rtc.get_system_clock_utc(msg.gps_time_h, msg.gps_time_m, msg.gps_time_s, ms);
+    }
+
+    msg.msl_altitude = loc.alt * 0.01 + 0.5;
+
+    send_packet((const uint8_t *)&msg, sizeof(msg));
+}
+
+/*
+  send a packet out
+ */
+void AP_Hott_Telem::send_packet(const uint8_t *b, uint8_t len)
+{
+    uint8_t crc = 0;
+    while (len) {
+        hal.scheduler->delay_microseconds(BYTE_DELAY_US);
+        if (uart->tx_pending()) {
+            continue;
+        }
+        len--;
+        crc += *b;
+        uart->write(*b++);
+    }
+    hal.scheduler->delay_microseconds(BYTE_DELAY_US);
+    uart->write(crc);
+}
+
+/*
+  thread to process requests
+ */
+void AP_Hott_Telem::loop(void)
+{
+    uart->begin(19200, 10, 10);
+    uart->set_unbuffered_writes(true);
+    uart->set_blocking_writes(true);
+
+    while (true) {
+        hal.scheduler->delay(1);
+
+        uint32_t n = uart->available();
+        if (n < 2) {
+            // wait for 2 bytes
+            continue;
+        }
+        if (n > 2) {
+            while (n--) {
+                uart->read();
+                hal.scheduler->delay_microseconds(20);
```

Comment:
This I don't understand.  We need to drain the uart slowly?!

Reply:
yes, this thread is high priority. If a bug leads us to get a huge number we don't want to chew all CPU draining 100M bytes